kiwmi/desktop/view.c

```diff
@@ -259,3 +284,191 @@ view_create(
+    child->mapped                  = true;
+    view_child_damage(child);
+}
+
+static void
+view_child_unmap_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, unmap);
+    child->mapped                  = false;
+    view_child_damage(child);
+}
+
+static void
+view_child_view_unmap_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, view_unmap);
+    child->mapped = false;
+    view_child_damage(child);
+}
+
+struct kiwmi_view_child *
+view_child_create(
+    struct kiwmi_view *view,
+    struct wlr_surface *wlr_surface,
+    enum kiwmi_view_child_type type,
+    const struct kiwmi_view_child_impl *impl)
+{
+    struct kiwmi_view_child *child = calloc(1, sizeof(*child));
+    if (!child) {
+        wlr_log(WLR_ERROR, "Failed to allocate view_child");
+        return NULL;
+    }
+
+    child->type        = type;
+    child->impl        = impl;
+    child->view        = view;
+    child->wlr_surface = wlr_surface;
+
+    wl_list_init(&child->link);
+    wl_list_init(&child->children);
+    child->mapped = false;
+
+    child->commit.notify = view_child_commit_notify;
+    wl_signal_add(&wlr_surface->events.commit, &child->commit);
+
+    child->surface_destroy.notify = view_child_surface_destroy_notify;
+    wl_signal_add(&wlr_surface->events.destroy, &child->surface_destroy);
+
+    child->new_subsurface.notify = view_child_new_subsurface_notify;
+    wl_signal_add(&wlr_surface->events.new_subsurface, &child->new_subsurface);
+
+    child->view_unmap.notify = view_child_view_unmap_notify;
+    wl_signal_add(&view->events.unmap, &child->view_unmap);
+
+    // Let the caller add the listeners to the respective events
+    child->map.notify = view_child_map_notify;
+    wl_list_init(&child->map.link);
+
+    child->unmap.notify = view_child_unmap_notify;
+    wl_list_init(&child->unmap.link);
+
+    wl_list_init(&child->new_popup.link);
+    wl_list_init(&child->owner_destroy.link);
+
+    view_init_subsurfaces(child, child->view);
+
+    return child;
+}
+
+struct kiwmi_view_child *
+view_child_subsurface_create(
+    struct kiwmi_view_child *parent,
+    struct kiwmi_view *view,
+    struct wlr_subsurface *subsurface)
+{
+    struct kiwmi_view_child *child = view_child_create(
+        view, subsurface->surface, KIWMI_VIEW_CHILD_SUBSURFACE, NULL);
+    if (child == NULL) {
+        return NULL;
+    }
+
+    child->wlr_subsurface = subsurface;
+
+    if (parent != NULL) {
+        child->parent = parent;
+        wl_list_insert(&parent->children, &child->link);
+    }
+
+    wl_signal_add(&subsurface->events.map, &child->map);
+    wl_signal_add(&subsurface->events.unmap, &child->unmap);
+
+    child->owner_destroy.notify = view_child_subsurface_owner_destroy_notify;
+    wl_signal_add(&subsurface->events.destroy, &child->owner_destroy);
+
+    child->mapped = true;
```

Comment:
This is what Sway does. I’ll check the specs.

Reply:
Specs:

> A sub-surface becomes mapped, when a non-NULL wl_buffer is applied and the parent surface is mapped. The order of which one happens first is irrelevant. A sub-surface is hidden if the parent becomes hidden, or if a NULL wl_buffer is applied. These rules apply recursively through the tree of surfaces.

It seems to be handled by wlroots, and when I removed it, Sublime Merge tooltips still work (I probably really need a well-behaved but edge-case client 😉).

Reply:
I’m dumb. Of course during my ‘test’ nothing was broken, because `child->mapped` is only read by the destructor, and damaging twice (on creation and destruction) should have the same effect as not damaging (in the edge case when the subsurface isn’t really used).